frontend/subrequests/lint/lint.go

```diff
@@ -0,0 +1,163 @@
+		Definition: sourceMap.Definition.ToPB(),
+		Data:       sourceMap.Data,
+	}
+	for i, source := range results.Sources {
+		if sourceEqual(source, newSource) {
+			return i
+		}
+	}
+	results.Sources = append(results.Sources, newSource)
+	return len(results.Sources) - 1
+}
+
+func (results *LintResults) AddWarning(rulename, description, url, fmtmsg string, sourceIndex int, location []parser.Range) {
+	sourceLocation := []*pb.Range{}
+	for _, loc := range location {
+		sourceLocation = append(sourceLocation, &pb.Range{
+			Start: pb.Position{
+				Line:      int32(loc.Start.Line),
+				Character: int32(loc.Start.Character),
+			},
+			End: pb.Position{
+				Line:      int32(loc.End.Line),
+				Character: int32(loc.End.Character),
+			},
+		})
+	}
+	pbLocation := pb.Location{
+		SourceIndex: int32(sourceIndex),
+		Ranges:      sourceLocation,
+	}
+	results.Warnings = append(results.Warnings, Warning{
+		RuleName:    rulename,
+		Description: description,
+		URL:         url,
+		Detail:      fmtmsg,
+		Location:    pbLocation,
+	})
+}
+
+func sourceEqual(a, b Source) bool {
+	if a.Filename != b.Filename || a.Language != b.Language {
+		return false
+	}
+	return bytes.Equal(a.Data, b.Data)
+}
+
+func (results *LintResults) ToResult() (*client.Result, error) {
+	res := client.NewResult()
+	dt, err := json.MarshalIndent(results, "", "  ")
+	if err != nil {
+		return nil, err
+	}
+	res.AddMeta("result.json", dt)
+
+	b := bytes.NewBuffer(nil)
+	if err := PrintLintViolations(dt, b); err != nil {
+		return nil, err
+	}
+	res.AddMeta("result.txt", b.Bytes())
+
+	res.AddMeta("version", []byte(SubrequestLintDefinition.Version))
+	return res, nil
+}
+
+func PrintLintViolations(dt []byte, w io.Writer) error {
+	var warnings LintResults
+
+	if err := json.Unmarshal(dt, &warnings); err != nil {
+		return err
+	}
+
+	// Here, we're grouping the warnings by rule name
+	lintWarnings := make(map[string][]Warning)
+	lintWarningRules := []string{}
+	for _, warning := range warnings.Warnings {
+		if _, ok := lintWarnings[warning.RuleName]; !ok {
+			lintWarningRules = append(lintWarningRules, warning.RuleName)
+			lintWarnings[warning.RuleName] = []Warning{}
+		}
+		lintWarnings[warning.RuleName] = append(lintWarnings[warning.RuleName], warning)
+	}
+	sort.Strings(lintWarningRules)
+
+	tw := tabwriter.NewWriter(w, 0, 0, 2, ' ', 0)
+	for _, rule := range lintWarningRules {
+		fmt.Fprintf(tw, "Lint Rule %s\n", rule)
+		for _, warning := range lintWarnings[rule] {
+			source := warnings.Sources[warning.Location.SourceIndex]
+			sourceData := bytes.Split(source.Data, []byte("\n"))
+			firstRange := warning.Location.Ranges[0]
+			if firstRange.Start.Line != firstRange.End.Line {
+				fmt.Fprintf(tw, "\t%s:%d-%d\n", source.Filename, firstRange.Start.Line, firstRange.End.Line)
+			} else {
+				fmt.Fprintf(tw, "\t%s:%d\n", source.Filename, firstRange.Start.Line)
+			}
+			fmt.Fprintf(tw, "\t%s\n", warning.Detail)
+			for _, r := range warning.Location.Ranges {
```

Comment:
Follow-up:

The source stacktrace doesn't seem to match the style in the progressbar:

```
Lint Rule FromAsCasing
  Dockerfile:1
  Lint Rule 'FromAsCasing': 'as' and 'FROM' keywords' casing do not match (line 1)
  1  |  FROM alpine as Base
```

```
 - Lint Rule 'FromAsCasing': 'as' and 'FROM' keywords' casing do not match (line 1)
The 'as' keyword should match the case of the 'from' keyword
Dockerfile:1
--------------------
   1 | >>> FROM alpine as Base
   2 |     RUN apk add git
   3 |     workdir /
--------------------
```

Reply:
Also, note that this description part is missing in print output. Is this intentional?

Reply:
It ended up being a lot of redundant information displayed and looked pretty cluttered. That being said, getting them to match the much superior progressbar style is an EXCELLENT idea. 😂